src/caffe/solver.cpp

```diff
@@ -266,10 +266,6 @@ void Solver<Dtype>::Step(int iters) {
     }
     ApplyUpdate();
 
-    // Increment the internal iter_ counter -- its value should always indicate
-    // the number of times the weights have been updated.
-    ++iter_;
-
```

Comment:
Why did this have to be moved to SGDSolver?

Reply:
Based on the comment: "its value should always indicate the number of times the weights have been updated."

Because now it is possible to update weights without having to call `step`, this has to be moved so that also calling directly `ApplyUpdate` increases it. I think that this is even cleaner from the design perspective. The code is next to the code which is updating weights. Not somewhere else.

Reply:
Ideally the base `Solver` would enforce the increment to `iter_`, but since all Caffe solvers are SGD solvers and other solvers that depart from this mold aren't expected I think the move is a pragmatic choice that avoids unnecessary interface code.